src/java/org/apache/cassandra/db/memtable/AbstractMemtable.java

```diff
@@ -0,0 +1,210 @@
+    protected AtomicLong minTimestamp = new AtomicLong(Long.MAX_VALUE);
+    // the write barrier for directing writes to this memtable or the next during a switch
+    protected volatile OpOrder.Barrier writeBarrier;
+    protected TableMetadataRef metadata;
+
+    public AbstractMemtable(TableMetadataRef metadataRef)
+    {
+        this.metadata = metadataRef;
+        this.columnsCollector = new ColumnsCollector(metadata.get().regularAndStaticColumns());
+    }
+
+    public void switchOut(OpOrder.Barrier writeBarrier, AtomicReference<CommitLogPosition> commitLogUpperBound)
+    {
+        assert this.writeBarrier == null;
+        this.writeBarrier = writeBarrier;
+        // This can prepare the memtable data for deletion; it will still be used while the flush is proceeding.
+        // A setDiscarded call will follow.
+    }
+
+    // decide if this memtable should take the write, or if it should go to the next memtable
+    public boolean accepts(OpOrder.Group opGroup, CommitLogPosition commitLogPosition)
+    {
+        // if the barrier hasn't been set yet, then this memtable is still taking ALL writes
+        OpOrder.Barrier barrier = this.writeBarrier;
+        if (barrier == null)
+            return true;
+        // if the barrier has been set, the write is for this memtable if it was issued before it
+        return barrier.isAfter(opGroup);
+    }
+
+    public TableMetadata metadata()
+    {
+        return metadata.get();
+    }
+
+    public long getLiveDataSize()
+    {
+        return liveDataSize.get();
+    }
+
+    public long getOperations()
+    {
+        return currentOperations.get();
+    }
+
+    public long getMinTimestamp()
+    {
+        return minTimestamp.get();
+    }
+
+    protected void updateMin(AtomicLong minTracker, long newValue)
+    {
+        while (true)
+        {
+            long memtableMinTimestamp = minTracker.get();
+            if (memtableMinTimestamp <= newValue)
+                break;
+            if (minTracker.compareAndSet(memtableMinTimestamp, newValue))
+                break;
+        }
+    }
+
+    RegularAndStaticColumns columns()
+    {
+        return columnsCollector.get();
+    }
+
+    EncodingStats encodingStats()
+    {
+        return statsCollector.get();
+    }
+
+    protected static class ColumnsCollector
+    {
+        private final HashMap<ColumnMetadata, AtomicBoolean> predefined = new HashMap<>();
+        private final ConcurrentSkipListSet<ColumnMetadata> extra = new ConcurrentSkipListSet<>();
+
+        ColumnsCollector(RegularAndStaticColumns columns)
+        {
+            for (ColumnMetadata def : columns.statics)
+                predefined.put(def, new AtomicBoolean());
+            for (ColumnMetadata def : columns.regulars)
+                predefined.put(def, new AtomicBoolean());
+        }
+
+        public void update(RegularAndStaticColumns columns)
+        {
+            for (ColumnMetadata s : columns.statics)
```

Comment:
Correctness wrt concurrency is not trivial in this class I think:

1. `update()` and `get()` may be interleaved, correct? (streaming case); if this is incorrect let's have explicit assertion that it doesn't happen   
2. if they can be interleaved then the `get()` caller may be a somewhat unexpected view of the columns (E.g. may see a newer update, but miss an older one). I _assume_ it is OK

Could we state 2 explicitly in some comment?